### ISSUE TITLE
Add a dynamic timeout window to BatchWithTimeout - BatchWithDynamicTimeout

### ISSUE THREAD

**Issue by bwagner5** [samber/lo#243]
It would be awesome if there was a dynamic window in the BatchWithTimeout func or a variant of that func where the Timeout would be the maximum time the batcher would wait to receive on the channel. Another IdleTimeout would engage if the receives slow to a certain rate.

For example:
```go
generator := func(yield func(int)) {
    for i := 0; i < 2; i++ {
        yield(i)
        time.Sleep(35*time.Millisecond)
    }
}

ch := lo.Generator(0, generator)

items1, length1, duration1, ok1 := lo.BatchWithDynamicTimeout(ch, 3,  10*time.Millisecond, 100*time.Millisecond)
// []int{1}, 1, 10ms, true

items2, length2, duration2, ok2 := lo.BatchWithDynamicTimeout(ch, 3,  10*time.Millisecond, 100*time.Millisecond)
// []int{2}, 1, 10ms, true


generator := func(yield func(int)) {
    for i := 0; i < 5; i++ {
        yield(i)
        time.Sleep(9*time.Millisecond)
    }
}

ch := lo.Generator(0, generator)

items3, length3, duration3, ok3 := lo.BatchWithDynamicTimeout(ch, 3,  10*time.Millisecond, 100*time.Millisecond)
// []int{1, 2, 3, 4, 5}, 5, 100ms, true
```

**Comment by samber** [comment]
Hi @bwagner5 

Can you provide a simple use case, please ?

**Comment by bwagner5** [comment]
Absolutely! The case that I had in mind was provisioning more capacity for a container orchestration system. The basic use-case is that containers come into the system that can't be scheduled to existing capacity, so you have to bin-pack them and provision new VMs.  

Obviously, putting 1 container per VM is not optimal which would be equivalent of receiving an event in the channel and immediately acting on it. But setting 1 timeout is also not efficient since you may wait the full timeout even if you only have 1 or 2 containers that come in. 

In this case, it's useful to have a dynamic window where you can react based on the flow of incoming events and stop early if the flow slows. This results in provisioning the VMs quicker when there's a signal that events are slowing down.  I think this works in other cases where you'd like to batch and flush early if there's indications of slowing events or some idle time indicating a logical batch can be handled. 

I work on a Kubernetes node autoscaling controller called Karpenter and we use this dynamic batching window in several places for efficiency: https://github.com/aws/karpenter/blob/main/pkg/controllers/provisioning/batcher.go
